src/main/java/uk/ac/ic/wlgitbridge/bridge/db/postgres/PostgresDBStore.java

```diff
@@ -0,0 +1,298 @@
+        "SELECT path "
+          + "FROM url_index_store "
+          + "WHERE project_name = ? "
+          + "AND url = ?"
+      );
+      statement.setString(1, projectName);
+      statement.setString(2, url);
+      ResultSet rs = statement.executeQuery();
+      if (rs.next()) {
+        String path = rs.getString(1);
+        rs.close();
+        statement.close();
+        return path;
+      } else {
+        return null;
+      }
+    } catch (Exception e) {
+      Log.error("Error while querying Postgres: {}", e.getMessage());
+      throw new RuntimeException(e);
+    }
+  };
+
+  @Override
+  public String getOldestUnswappedProject() {
+    try {
+      Statement statement = connection.createStatement();
+      ResultSet rs = statement.executeQuery(
+        "SELECT name FROM projects" +
+          "  ORDER BY last_accessed ASC" +
+          "  LIMIT 1;\n"
+      );
+      if (rs.next()) {
+        String project = rs.getString(1);
+        rs.close();
+        statement.close();
+        return project;
+      } else {
+        return null;
+      }
+    } catch (Exception e) {
+      Log.error("Error while querying Postgres: {}", e.getMessage());
+      throw new RuntimeException(e);
+    }
+  };
+
+  @Override
+  public int getNumUnswappedProjects() {
+    try {
+      Statement statement = connection.createStatement();
+      ResultSet rs = statement.executeQuery(
+        "SELECT COUNT(*)\n" +
+          "    FROM projects\n" +
+          "    WHERE last_accessed IS NOT NULL"
+      );
+      if (rs.next()) {
+        int n = rs.getInt(1);
+        rs.close();
+        statement.close();
+        return n;
+      } else {
+        return 0;
+      }
+    } catch (Exception e) {
+      Log.error("Error while querying Postgres: {}", e.getMessage());
+      throw new RuntimeException(e);
+    }
+  };
+
+  @Override
+  public ProjectState getProjectState(String projectName) {
+    try {
+      PreparedStatement statement = connection.prepareStatement(
+        "SELECT last_accessed\n" +
+          "    FROM projects\n" +
+          "    WHERE name = ?;"
+      );
+      statement.setString(1, projectName);
+      ResultSet rs = statement.executeQuery();
+      ProjectState result;
+      if (rs.next()) {
+        if (rs.getTimestamp(1) == null) {
+          result = ProjectState.SWAPPED;
+        } else {
+          result = ProjectState.PRESENT;
+        }
+      } else {
+        result = ProjectState.NOT_PRESENT;
+      }
+      rs.close();
+      statement.close();
+      return result;
+    } catch (Exception e) {
+      Log.error("Error while querying Postgres: {}", e.getMessage());
+      throw new RuntimeException(e);
+    }
+  };
+
+  @Override
+  public void setLastAccessedTime(String projectName, Timestamp time) {
+    try {
+      PreparedStatement statement = connection.prepareStatement(
+        "UPDATE projects\n" +
+          "SET last_accessed = ?\n" +
+          "WHERE name = ?"
```

Comment:
Seems like there is inconsistency in this file on whether SQL statements are terminated with a semicolon or not. I don't suppose it matters a great deal though?

Reply:
I think this is a side-effect of adapting the original sqlite-flavoured sql code. Will update to use semi-colons consistently :+1: